qiskit_aer/quantum_info/states/aer_statevector.py

```diff
@@ -184,27 +184,56 @@ def _aer_evolve_circuit(aer_state, circuit, qubits):
     @staticmethod
     def _aer_evolve_instruction(aer_state, inst, qubits):
         """Apply instruction into aer_state"""
+
         params = inst.params
+
+        def _assert_qubits(expected):
+            if len(qubits) != expected:
+                raise AerError(
+                    f"Cannot apply a {expected}-qubit instruction {inst} to qubits {qubits}"
+                )
+
+        def _assert_params(expected):
+            if len(params) < expected:
+                raise AerError(f"Need {expected} parameters, but actually {len(params)}")
+
         if inst.name in ['u', 'u3']:
-            aer_state.apply_mcu(qubits[0:len(qubits) - 1], qubits[len(qubits) - 1],
-                                params[0], params[1], params[2])
-        elif inst.name in ['x', 'cx', 'ccx']:
-            aer_state.apply_mcx(qubits[0:len(qubits) - 1], qubits[len(qubits) - 1])
-        elif inst.name in ['y', 'cy']:
-            aer_state.apply_mcy(qubits[0:len(qubits) - 1], qubits[len(qubits) - 1])
-        elif inst.name in ['z', 'cz']:
-            aer_state.apply_mcz(qubits[0:len(qubits) - 1], qubits[len(qubits) - 1])
+            _assert_qubits(1)
+            _assert_params(3)
+            aer_state.apply_u(qubits[0], params[0], params[1], params[2])
```

Comment:
If you're unpacking the qubits and parameters, you probably don't need to waste time doing the assertions - you'll just get an `IndexError` if they're out of bounds anyway, which I think is probably fine for something that's meant to be a bit higher performance?

It's up to you, really - I feel like we could go either way.

Reply:
Maybe `IndexError` will be consistent with the other parts. I would like to take `_assert*`.